Transform

sonals opened the pull request:
<!-- Thanks for sending a pull request! Please fill out below, remove sections that don't apply for your pull request.  -->
#### Problem solved by the commit

1. New infrastructure which enables writing aie2p ctrlcode transforms using a pass manager.  p
2. Added a new utility, aiebu-transform to run the transforms

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
New feature

#### How problem was solved, alternative solutions (if any) and why they were rejected
Added a new pass manager and a base class for defining a pass.  Currently the pass manager runs a canned sequence of transforms. However, in the final solution a JSON file will determine which transforms to run. Note that the current ELF generated by aiebu cannot be read back and rewritten by aiebu. This issue needs to be solved first before the transforms can be enabled and saved.

#### Risks (if any) associated the changes in the commit
NA

#### What has been tested and how, request additional testing if necessary
Compile only tested

#### Documentation impact (if any)

stsoe commented on src/cpp/analyzer/passmanager.h:
```diff
@@ -0,0 +1,53 @@
+    // Currently we are running a precannded sequence of transforms.
+    ELFIO::Elf_Half sec_num = m_elf.sections.size();
+    for ( int i = 0; i < sec_num; ++i ) {
+      ELFIO::section *psec = m_elf.sections[i];
```
Range based?

```suggestion
  for (auto psec : m_elf_sections) {
```

sonals replied:
Done!

stsoe commented on src/cpp/utils/transform/CMakeLists.txt:
```diff
@@ -0,0 +1,63 @@
+  ${AIEBU_SOURCE_DIR}/src/cpp/include
+  ${AIEBU_SOURCE_DIR}/src/cpp/cxxopts/include
+  ${AIEBU_SOURCE_DIR}/src/cpp
+  ${AIEBU_SOURCE_DIR}/src/cpp/ELFIO
+  ${AIEBU_SOURCE_DIR}/src/cpp/disassembler
+  ${AIEBU_SOURCE_DIR}/src/cpp/common
+  ${AIEBU_SOURCE_DIR}/src/cpp/ops
+  ${AIEBU_SOURCE_DIR}/src/cpp/preprocessor
+  ${AIEBU_SOURCE_DIR}/
+  ${Boost_INCLUDE_DIRS}
```
PLease don't continue a bad trend.   Can we `#include` relative to say `${AIEBU_SOURCE_DIR}/src/cpp` ?   This makes it much easier for code readers to navigate to included file.

sonals replied:
I have removed unnecessary paths.

stsoe commented on src/cpp/analyzer/basicpass.h:
```diff
@@ -0,0 +1,60 @@
+  added     // a new node has been created and added to the ctrlcode list
+};
+
+template <typename aie2p_type> struct basic_node {
```
```suggestion
template <typename aie2p_type> 
struct basic_node {
```

larry9523 commented on src/cpp/utils/transform/transform.cpp:
```diff
@@ -0,0 +1,99 @@
+  ebin.load(result["filename"].as<std::string>());
+  // We fail in the save even without any transforms. First this needs to be
+  // debugged and resolved before we can save any transformed ctrlcode. Once
+  // this bug is fixed commen out the ebi.save() from here.
```
@donwalkarsoham is trying to fix it by adding section sh_addr and segment p_vaddr by its size and alignment in ELF. Those apply to "PROGBITS" and "DYNAMIC" sections.

donwalkarsoham replied:
Added PR https://github.com/Xilinx/aiebu/pull/149/files to fix issue from AIEBU side to avoid changes in ELFIO.
In aiebu we add virtual address of loadable segments so ELFIO::save() is able to layout the order of the segment sections correctly.

sonals commented:
Added a pre-transform operation which does the following:

1. Upgrades legacy ELF to add addresses to sections and segments -- the addresses match the physical offset
2. Remove ALLOC flag from all sections except text and data
3. Drops the  DYNAMIC segment. .dynamic section is enough
4. Changed the machine code to the official code, 269, allocated by ELF standards body

sonals commented on src/cpp/analyzer/basicpass.h:
```diff
@@ -0,0 +1,60 @@
+  added     // a new node has been created and added to the ctrlcode list
+};
+
+template <typename aie2p_type> struct basic_node {
```
Updated!

sonals commented on src/cpp/analyzer/basicpass.cpp:
```diff
@@ -0,0 +1,392 @@
+    : m_nodes(nodes), m_pdiid(pdiid), m_pdisize(pdisize) {}
+
+  void transform() override {
+    XAie_LoadPdiHdr *load = new XAie_LoadPdiHdr();
```
Done!
